### DISCUSSION TITLE
Can't display the Date an order was placed

### DISCUSSION THREAD

**Discussion by wautersj** [duncanmcclean/simple-commerce#947] [Help] [answered]
There seems to be no way to display the date in the Orders overview OR the order detail.

In context of a webshop, this feels important. 
In any case, it is important to our customers (for whom we build shops) and for ourselves (our own shop).

I've done a search for similar discussions/issues: 
https://github.com/duncanmcclean/simple-commerce/discussions/896

To me, it feels like a missing feature; and would be of great value to have it in SC.

![Screenshot 2023-12-18 at 10 39 24](https://github.com/duncanmcclean/simple-commerce/assets/4070167/b58a2836-7bad-442d-8239-62eb9d02b054)

**Comment by duncanmcclean** [comment]
Hey 👋

Yes, I know. It was accidentally removed after some changes to the way order/payment statuses work in v5 but I'm planning to add it back in v6 (which I'm planning to release Q1 next year), see https://github.com/duncanmcclean/simple-commerce/discussions/896.

In the meantime, you can add it using a [Computed Field](https://statamic.dev/computed-values#overview). 

1. First, edit the "Orders" blueprint and add an "Order Date" field. Make sure "Visibility" is set to "Computed" and "Listable" is set to "Shown by default"
2. Secondly, add the following code to the `boot` method in your `AppServiceProvider`:

```php
// app/Providers/AppServiceProvider.php

public function boot()
{
    // ...

    \Statamic\Facades\Collection::computed('orders', 'order_date', function ($entry, $value) {
        return \Illuminate\Support\Arr::get($entry->get('status_log'), 'placed');
    });
}
```

3. You should now be able to select the "Order Date" column from the column picker and you'll see the date appear.

**Comment by wautersj** [comment]
I think we're almost there.But am missing something...

Btw, I saw that when editing the blueprint via the Statamic web UI; it adds the field to an other orders.yaml. In which case the field was not found and could not create an extra column. When adding that field data to the actual blueprint, it found the field, and after adding the boot code, the field became listed in de columns list.

<img width="480" alt="Screenshot 2023-12-18 at 11 54 13" src="https://github.com/duncanmcclean/simple-commerce/assets/4070167/a76fc560-9c26-4692-8816-ed2658b29413">

--

But now the column is added, no value is displayed for the orders. Am i still missing a crucial step?
(I tried a DD in the methode but I can't get it to be triggered when visiting any cp route)

![Screenshot 2023-12-18 at 11 53 02](https://github.com/duncanmcclean/simple-commerce/assets/4070167/bb762530-4bbd-4643-b2fc-ca017083ba2e)
<img width="1308" alt="Screenshot 2023-12-18 at 11 56 10" src="https://github.com/duncanmcclean/simple-commerce/assets/4070167/521ea197-723d-4ad7-9dee-c90bce66d3d9">

**Reply by duncanmcclean** [reply]
> Btw, I saw that when editing the blueprint via the Statamic web UI; it adds the field to an other orders.yaml. In which case the field was not found and could not create an extra column. When adding that field data to the actual blueprint, it found the field, and after adding the boot code, the field became listed in de columns list.

I'm not quite sure I fully understand what you mean but it sounds weird it's generating another blueprint file 🤔 

> But now the column is added, no value is displayed for the orders. Am i still missing a crucial step?

Are `orders` and `order_date` correct? (`orders` should be the handle of your orders collection and `order_date` should be the name of the field).

Is the field's "Visibility" set to "Computed"?

**Comment by wautersj** [comment]
If I edit the blueprint of 'orders' via the CP; 
then it edits the file located at: 'blueprints/collections/orders/order.yaml'
and then, in combination with the extra code in the AppServiceProvider, I don't get to select the Order Date in de columns (it doesn't show up)

If i edit the blueprint of 'orders' by editing the file located at 'blueprints/order.yaml'
and then, in combination with the extra code in the AppServiceProvider, I do get to select the Order Date in de columns (then i shows up)

<img width="316" alt="Screenshot 2023-12-18 at 13 02 05" src="https://github.com/duncanmcclean/simple-commerce/assets/4070167/ee7042ac-f47a-4094-b030-e9419dcc1389">

![Screenshot 2023-12-18 at 13 05 27](https://github.com/duncanmcclean/simple-commerce/assets/4070167/41865966-3ca3-4649-9022-99919a7cbc90)

<img width="1207" alt="Screenshot 2023-12-18 at 13 08 36" src="https://github.com/duncanmcclean/simple-commerce/assets/4070167/c1a35121-e8b4-447d-83f2-49af13ee8106">

Yes, the visibility is computed.
This field is listed in the blueprints/order.yaml file.

<img width="987" alt="Screenshot 2023-12-18 at 13 10 14" src="https://github.com/duncanmcclean/simple-commerce/assets/4070167/2e40a334-4fbe-46fe-b736-93261b6f8f57">

**Reply by duncanmcclean** [reply, marked as answer]
Oh, sorry, didn't realise you were storing orders in a database, in which case editing the `resources/order.yaml` blueprint is correct. You can delete the "Orders" collection and its blueprint since you're not using entries.  

However, instead of the computed code in the `AppServiceProvider`, you'll need to override the Order Eloquent model and create a model accessor to return the order date.

1. Create a new file in `App\Models` called `Order.php`, with the following contents:

```php
<?php

namespace App\Models;

use DoubleThreeDigital\SimpleCommerce\Orders\OrderModel as BaseOrder;
use Illuminate\Database\Eloquent\Casts\Attribute;
use Illuminate\Support\Arr;

class Order extends BaseOrder
{
    public function orderDate(): Attribute
    {
        return Attribute::make(
            get: function () {
                if (! Arr::has($this->data, 'status_log')) {
                    return null;
                }

                return Arr::get($this->data, 'status_log.placed');
            },
        );
    }
}
```

2. In your Simple Commerce config file (`config/simple-commerce.php`), change the `model` from `\DoubleThreeDigital\SimpleCommerce\Orders\OrderModel::class` to `\App\Models\Order::class`
3. Visit your listing table and hope for the best 🎉 

_^ I haven't manually tested these steps as I don't have the time at the minute but I think they should work._

**Comment by wautersj** [comment]
That were the right steps in the right direction, thanks!

Btw, I noticed that I also had to switch the model referenced in runway.php
<img width="834" alt="Screenshot 2023-12-18 at 15 03 07" src="https://github.com/duncanmcclean/simple-commerce/assets/4070167/16403cbf-306c-4546-991c-791d8a35ff8a">
That's OK right?

**Reply by duncanmcclean** [reply]
Ah yes, you need to change it there too, well spotted!

**Comment by duncanmcclean** [comment]
I've been making some changes to how the "status log" feature works and in the process, I've added an "Order Date" field to orders so you don't need to do it yourself.

After updating to [v5.7.0](https://github.com/duncanmcclean/simple-commerce/releases/tag/v5.7.0), you can follow my previous set of instructions in reverse. (get rid of the custom Eloquent model, switch back to the "base" Simple Commerce one).

The only thing you'll want to keep is the `order_date` field on your blueprint. Other than that, it should all "just work" now. 🚀
